Source/WebCore/platform/graphics/NativeImage.cpp

```diff
@@ -104,11 +104,45 @@ DestinationColorSpace NativeImage::colorSpace() const
     return m_backend->colorSpace();
 }
 
+bool NativeImage::hasHDRContent() const
+{
+    return colorSpace().usesITUR_2100TF();
+}
+
 Headroom NativeImage::headroom() const
 {
     return m_backend->headroom();
 }
 
+void NativeImage::drawToneMapping(GraphicsContext& context, const FloatRect& destinationRect, const FloatRect& sourceRect, ImagePaintingOptions options)
+{
+    auto colorSpaceForToneMapping = [](GraphicsContext& context) {
+#if PLATFORM(IOS_FAMILY)
+        // iOS typically renders into extended range sRGB to preserve wide gamut colors, but we want
+        // a non-extended range colorspace here so that the contents are tone mapped to SDR range.
+        UNUSED_PARAM(context);
+        return DestinationColorSpace::DisplayP3();
+#else
+        // Otherwise, match the colorSpace of the GraphicsContext.
+        return context.colorSpace();
```

Comment:
If we use the destination context's colorspace on macOS, it would seem that using the intermediate buffer doesn't introduce a behavior change. Is it just impacting behavior because it's Unaccelerated? If so, that deserves a command and a radar link.

Reply:
Also, here in cross-platform code doesn't seem to be the right place to work around Apple platform limitations.

Reply:
Although the destination color space on macOS may extended dynamic range colorspace, the pixel format of the  intermediate buffer is BGRA8. This will force the tone-mapping anyway.

Why is this function is cross-platform? Because all platforms can save memory if `dynamic-range-limit: standard;`. Like what we do for Apple ports, they can create normal 4-bytes-per-pixel buffers for the layers. So the tone-mapping will be needed in this case.